Integrate Parity Signer into Fether

axelchalon opened the pull request:
Integrate Parity Signer (closes #146)
- Add option to import a Parity Signer account. It is then stored in local storage.
- Send transactions with Parity Signer accounts
- Refactor sendStore so that "token" (and chainId) is inside this.tx; avoids having to pass `token` around every function, and it makes sense to put it in the object since it is part of the tx info.
- Add account type (either node or signer) to accountsInfo
- Add account type to withAccount; output props are now `{account: {address, type}}` instead of `{accountAddress}`

For some reason some of the changes I introduced makes TxForm (and seemingly AccountsList as well) reset/remount on every block; not sure why that is, need to track this down. Makes it difficult to send transactions.

axelchalon commented on packages/fether-react/src/utils/withAccount.js:
```diff
@@ -5,18 +5,45 @@
+      <AccountAddressFromRouter>
+        {accountAddress => {
+          const DecoratedComponent = compose(
+            light({
```
```js
            light({
              transactionCount: () =>
                transactionCountOf$(accountAddress).pipe(withoutLoading())
            }),
```
This makes the component remount on every new block (blank screen for a short period then TxForm appears again, reset). Not sure how to rewrite this?
If I remove those lines the TxForm fields just reset on every block (even though there is no remounting)

amaurymartiny replied:
I believ HOCs in the render method is an anti-pattern.

Would something like this work?
```javascript
const WithAccount = compose(
  withRouter,
  withAccountsInfo,
  light({
    transactionCount: props =>
      transactionCountOf$(props.router.accountAddress).pipe(withoutLoading())
  }),
  mapProps(...)
)(props => props.children(props.account));

export default Component => initialProps => (
  <WithAccount>
    {account => <Component {...initialProps} account={account} />}
  </WithAccount>
);
```

axelchalon replied:
it does work! much cleaner :+1: still doesn't resolve the reset issue though, I'll continue investigating

axelchalon replied:
actually my bad, it does seem to have resolved the issue!! :tada:

Tbaut commented:
Awesome!! I gave it a go and here is an early feedback:
- I opened the scanner view to import an account, it took me a minute or so to grab my phone and bring up the Parity Signer QR code, in the mean time, Fether switched to the "[new account](https://pix.toile-libre.org/upload/original/1546865199.jpg)" view without me doing anything (note that I could reproduce it another time, but it doesn't happen every time), that's not critical. 

- The left part of the webcam scanner is grey (here is my webcam with a postit on it) the gray column on the left is not expected I guess.
![image](https://user-images.githubusercontent.com/33178835/50768760-e7c46600-1281-11e9-8def-8207dd49932e.png)

- The confirm button is not the default one when entering name (similar to https://github.com/paritytech/fether/issues/280)

- We don't need to have a password for such an account as the security is handled by Parity Signer itself.

- When clicking on the account balance to send some ETH I got a  
`TypeError: this.project is not a function` in switchMap.js:55
![image](https://user-images.githubusercontent.com/33178835/50768547-1aba2a00-1281-11e9-920c-720d967bf1a5.png)

axelchalon commented:
> The left part of the webcam scanner is grey (here is my webcam with a postit on it) the gray column on the left is not expected I guess.

Okay, I have the same thing. Good to know it's not just me :P Might be an issue in `@parity/qr-signer`

> The confirm button is not the default one when entering name (similar to #280)

Not sure I get it. When I'm importing a Parity Signer account, after scanning I land on the AccountName screen with the account name field automatically focused ; once I've entered something, if I press enter it submits the form as expected

> We don't need to have a password for such an account as the security is handled by Parity Signer itself.

Cannot reproduce; after entering the account name for a Parity Signer account I'm redirected to the accounts list.

> When clicking on any account balance to send some ETH I get a TypeError: this.project is not a function in switchMap.js:55

Cannot reproduce. Maybe reinstall the dependencies?

Tbaut commented:
Yup, most of the problems above where because of me, I updated the post above. Here is what I found (beside the refresh problem):

- The QR code to scan the transaction isn't in a card (with white backgrounds) and also not centerd
![image](https://user-images.githubusercontent.com/33178835/50783858-40a8f400-12ac-11e9-9763-805e7a954877.png)
- Same for the Webcam scanner (this time the column on the left is black)
![image](https://user-images.githubusercontent.com/33178835/50783892-5f0eef80-12ac-11e9-9b7c-59f2adbf983f.png)

Other than that it's working great!!

ltfschoen commented:
I tried using the new features using a Samsung Galaxy S8 Note mobile using Parity Signer 2.0 Beta and on macOS Mojave 10.14.

1) If we create an account just using Fether (without Parity Signer), then once its created nothing is in localStorage until you associate a token (other than ETH) like THIBCoin with the account it's stored in localStorage as:
  * Key: localforage/__paritylight::tokens::0x005072Fb985cc64d8A1Ce00ECe26279f73c12F8a::kovan
  * Value: {"ETH":{"address":"ETH","decimals":18,"logo":"data:image/png;base64,iVB...II=","name":"Ether","symbol":"ETH"},"0x4A6e6C3868A279e1D9047B42C3fB356FF4680003":{"address":"0x4A6e6C3868A279e1D9047B42C3fB356FF4680003","decimals":18,"name":"THIBCoin","symbol":"TIB","logo":"http://pix.toile-libre.org/upload/original/1525765045.jpg"}}

Note: If we import an account that was created with Parity Signer that is from a different network other than Kovan, it just shows the Parity Signer logo inside their identicon, could we also show what network it is associated with
just like we do in Parity Signer and colour-code it?

2) After importing from Parity Signer it's stored in localStorage as the following. Notice how I've had to name the accounts with the network name they were created on?:
  * Key: localforage/__paritylight::paritySignerAccounts
  * Value: [
    {"address":"0x006e27b6a72e1f34c626762f3c4761547aff1421","name":"test-signer-kovan"},
    {"address":"0x000ec8c79dac74d13b7e4743ae52f78d7186ba6d","name":"test-signer-kovan"},
    {"address":"0x00d7bb9dbd761fdf00d78f74710ad6bdcaed8d48","name":"test-signer-mainnet"}
  ]

<img width="360" alt="screen shot 2019-01-10 at 5 30 38 pm" src="https://user-images.githubusercontent.com/6226175/50950496-40079c00-14fe-11e9-9c14-6cb496bf658f.png">

3) After importing an account from Parity Signer it:
  * Resets the value stored under the following key (which may have contained information about a different account, see Step 4) to:
    * Key: _localforage_sys/_localforage_observable_sys
    * Value: {"name":"localforage","storeName":"keyvaluepairs","key":"__paritylight::paritySignerAccounts","methodName":"setItem","valueChange":true,"success":true,"ticks":1547094958181}

  * Note: It allows us to import an account that we previous imported again, but it appears to do so correctly. Should we even allow the user to do this anyway? Or on the "Import Account" page should we show an error that the "Account is already in the account list"?

  * Note: It scans the "account" QR code really quickly and even when I aim it at 45 degree angle. Whereas later on when I scan the "signature" QR code I had to aim it properly

  * Note: I never saw the gray column on the "Import account" page where I scanned my QR code, which @Tbaut experienced. This is what I see (notice how I hid away just before taking the screenshot)

<img width="360" alt="screen shot 2019-01-10 at 5 25 01 pm" src="https://user-images.githubusercontent.com/6226175/50950455-18b0cf00-14fe-11e9-91f6-64962e3ec9f7.png">

4) If we then associate a token like THIBCoin with the account that we imported from Parity Signer:
  * Overwrites the values stored under the following key in localStorage to the following:
    * Key: _localforage_sys/_localforage_observable_sys
    * Value: {"name":"localforage","storeName":"keyvaluepairs","key":"__paritylight::tokens::0x00ca490242abb36cfad5697bafa69fbd1824389d::kovan","methodName":"setItem","valueChange":true,"success":true,"ticks":1547095163503}

    * Note: If I add a token (other than ETH) like THIBCoin to an account that I created using Parity Signer a network other than Kovan (i.e. Ethereum Mainnet), then it still adds it as shown above, but appears to associate the tokens with Kovan Network since it has `::kovan`, is this correct?

  * Adds the following key/value to localStorage. If we import and account that we previously imported it retains the tokens associated correctly:

    * Key: localforage/__paritylight::tokens::0x005072Fb985cc64d8A1Ce00ECe26279f73c12F8a::kovan
    * Value: {"ETH":{"address":"ETH","decimals":18,"logo":"data:image/png;base64,iV...II=","name":"Ether","symbol":"ETH"},"0x4A6e6C3868A279e1D9047B42C3fB356FF4680003":{"address":"0x4A6e6C3868A279e1D9047B42C3fB356FF4680003","decimals":18,"name":"THIBCoin","symbol":"TIB","logo":"http://pix.toile-libre.org/upload/original/1525765045.jpg"}}

    * Note: The address shown as `0x4A6e6C3868A279e1D9047B42C3fB356FF4680003` is the Contract Address of the THIBCoin token

5) If we remove the token THIBCoin from an account that we imported from Parity Signer, it:
  * Overwrites 
    * Key: _localforage_sys/_localforage_observable_sys
    * Value: {"name":"localforage","storeName":"keyvaluepairs","key":"__paritylight::tokens::0x00d7bb9dbd761fdf00d78f74710ad6bdcaed8d48::kovan","methodName":"setItem","valueChange":true,"success":true,"ticks":1547096914039}

  * Updates the following in localStorage:
    * Key: localforage/__paritylight::tokens::0x00d7bb9dbd761fdf00d78f74710ad6bdcaed8d48::kovan
    * Value: {"ETH":{"address":"ETH","decimals":18,"logo":"data:image/png;base64,iVB...II=","name":"Ether","symbol":"ETH"}}

6) After importing an an account from Parity Signer that was created on a network (tested so far using Ethereum mainnet, and Kovan). If within Fether I try to "Import Account" and enter the recovery words that were shown in Parity Signer, it returns "The passphrase was not recognised... ". Shouldn't we be able to recover an account using the keyphrase both from Parity Signer and Fether? 

7) At one stage after I was testing to see if I could recover a Parity Signer keyphrase using Fether, when I then went to scan a QR code to import an account it wasn't working, and when I checked the console it displayed error:
```
6b468d40-e396-45c6-aede-2c878f3e578a:1 Uncaught RangeError: Invalid typed array length: Infinity
    at typedArrayConstructByLength (<anonymous>)
    at new Uint8ClampedArray (native)
    at Function.o.createEmpty (blob:http://localhost:3000/6b468d40-e396-45c6-aede-2c878f3e578a:1:917)
    at Object.a [as extract] (blob:http://localhost:3000/6b468d40-e396-45c6-aede-2c878f3e578a:4:26907)
    at t (blob:http://localhost:3000/6b468d40-e396-45c6-aede-2c878f3e578a:1:4753)
    at blob:http://localhost:3000/6b468d40-e396-45c6-aede-2c878f3e578a:5:744
o.createEmpty @ 6b468d40-e396-45c6-aede-2c878f3e578a:1
a @ 6b468d40-e396-45c6-aede-2c878f3e578a:4
t @ 6b468d40-e396-45c6-aede-2c878f3e578a:1
(anonymous) @ 6b468d40-e396-45c6-aede-2c878f3e578a:5

```

8) After transferring some ETH to the Parity Signer account that I imported into Fether. I then tried to see what would happen if I transferred some of that ETH from an account that was created using Parity Signer on the Kovan network to another account (I tried sending it to another account that was created using Parity Signer on the Ropsten network, and then tried sending it to an account that was created using Fether on the Kovan network, but both exhibited the same behaviour as follows). It showed "Scan" instead of "Send" correctly, and instead of prompting me for a password (as it does for accounts created in Fether), it correctly asks me to "Please scan the QR code of the transaction on Parity Signer". So I scan the QR code that's shown with Parity Signer, and click the "Sign" button within Parity Signer, and then Parity Signer displays another QR code and prompts me to "SCAN SIGNATURE", so in Fether I click the "Next" button and on the next screen I use my webcam to scan the QR code that's shown in Parity Signer. But then it just goes back to the "Send Ether" page with the amount and recipient values still entered into the input fields but the fields aren't editable, and nothing is sent when I check the account on BlockScout, and nothing is shown in the console to indicate any error.

<img width="364" alt="screen shot 2019-01-10 at 5 04 48 pm" src="https://user-images.githubusercontent.com/6226175/50950417-f6b74c80-14fd-11e9-88e4-bbd1ddd3da04.png">

Note: I think the QR code that's shown needs some padding and should be centred on the screen with CSS

9) I think this is unrelated to this PR, but on the account page, the gap between the identicon and the account name is too small in my opinion

<img width="332" alt="screen shot 2019-01-10 at 4 53 01 pm" src="https://user-images.githubusercontent.com/6226175/50950397-e99a5d80-14fd-11e9-8ce7-f1298111972f.png">

Very impressive work! So far I've only performed "user testing" and haven't yet reviewed the code, apart from having a quick peek :). I'll look at it tonight

Tbaut commented:
1. > could we also show what network it is associated with just like we do in Parity Signer and colour-code it?

This will be needed soon (if we have some network change ability) but I think it's outside of the scope of this PR. (I want to release 0.2 as soon as this PR is merged, let's keep it simple).

2. >It allows us to import an account that we previous imported again, but it appears to do so correctly. Should we even allow the user to do this anyway? 

Good catch. Probably not if we want to be consistent.

2. >This is what I see

Too much sun/light visible on the picture, my (Berlin calibrated) eyes couldn't process it :cloud_with_rain: 

2. >It scans the "account" QR code really quickly and even when I aim it at 45 degree angle. Whereas later on when I scan the "signature" QR code I had to aim it properly

The signature contains much more info, it is a much more complex QR code that require better light and overall scanning conditions.

5. >Shouldn't we be able to recover an account using the keyphrase both from Parity Signer and Fether?

Damn, that's clearly a bug in Parity Signer, for some reason it generates a 11 words recovery. This must be fixed on Signer. For those who created an account with 11 words though it's be nice to support it in Fether. I understand if you're reluctant to do this though.

6. > transferred some of that ETH from an account that was created using Parity Signer on the Kovan network to another account (I tried sending it to another account that was created using Parity Signer on the Ropsten network

This is an edge case that I propose to handle separately. For the beta let's assume ppl work on the same network.

amaurymartiny commented on packages/fether-react/src/stores/sendStore.js:
```diff
@@ -75,6 +87,41 @@ export class SendStore {
+   */
+  @action
+  signRaw = signature => {
+    const signed = signTransactionWithSignature(this.tx, signature);
```
Hmm, so `postRaw$('0x' + signature)` won't work?

axelchalon replied:
hmm, not sure I understand
signature is just the signature, postRaw$ needs to be provided with a signed tx (i.e. tx+sig)
https://ethereum.stackexchange.com/a/2097
to my understanding: signature is serialized v, s, r; postRaw$ needs serialized nonce, gasprice, startgas, to, value, data, v, r, s

amaurymartiny replied:
Ah, right, I thought Parity Signer did the rlp(tx + signature) part, my bad!

amaurymartiny commented on packages/fether-react/src/stores/createAccountStore.js:
```diff
@@ -171,6 +192,17 @@ export class CreateAccountStore {
+  };
+
+  // Returns true for Signer account imports
+  // @computed get
```
Shouldn't be commented?

amaurymartiny commented on packages/fether-react/src/stores/createAccountStore.js:
```diff
@@ -8,12 +8,16 @@ import { action, observable } from 'mobx';
 import hdkey from 'ethereumjs-wallet/hdkey';
 
+import LS_PREFIX from './utils/lsPrefix';
+import localForage from 'localforage';
```
import order

ltfschoen replied:
LGTM

amaurymartiny commented on packages/fether-react/src/Send/SignedTxSummary/SignedTxSummary.js:
```diff
@@ -0,0 +1,135 @@
+              <TokenBalance
+                decimals={6}
+                drawers={[
+                  <Form
```
It would make sense to factorize code between this and TxForm, if it's easily doable

amaurymartiny commented on packages/fether-react/src/Accounts/CreateAccount/AccountImportOptions/AccountImportOptions.js:
```diff
@@ -67,51 +70,97 @@ class AccountImportOptions extends Component {
     }
   };
 
+  handleSignerImported = async ({ address }) => {
```
This component is starting to do a lot of different stuff, it would be good to split it out. Not in the scope of this PR though, probably with #264.

ltfschoen commented:
## 5) Importing Account in Fether using the passphrase that was generated when I created an in Parity Signer **(Continued)**

Since my previous comment https://github.com/paritytech/fether/pull/336#issuecomment-452986828, @Tbaut created this PR https://github.com/paritytech/fether/pull/351 so that if we created an account on Parity Signer we could import it into Fether using its 11 Recovery Words, and his fix works!

But I encountered a problem, which I'll now explain...

So I'd previously created a Parity Signer account and imported it into Fether using the QR scanner, and it displays successfully in my account list (with the Parity Signer logo embedded in the identicon). I had topped it up with some ETH and THIBCoin.

In my previous comment https://github.com/paritytech/fether/pull/336#issuecomment-452986828 I tried importing the same Parity Signer account into Fether that I had already previously imported into Fether from Parity Signer and was already shown in my account list, and found that it didn't prevent me from doing so (even though I expected it to say `Account 0x... already listed`), so I think we need to apply to "Recover from Parity Signer" the changes that were made in this PR for the other ways to import https://github.com/paritytech/fether/pull/327, but it also didn't cause any issues.

But now I wanted to see what would happen if I tried to import a Parity Signer account using "Recover from Seed Phrase" in Fether, even though I'd already previously imported it using "Recover from Parity Signer" and the account was already in my account list in Fether under the name "test-signer2", as shown below:

<img width="351" alt="screen shot 2019-01-11 at 10 04 29 am" src="https://user-images.githubusercontent.com/6226175/51002716-4f332c00-1588-11e9-8286-d36506a09ce3.png">

So I entered the recovery words that were generated when I created my Parity Signer account, but instead of it displaying `Account 0x... already listed` it took me to the "Please give this account a name:" page as shown below. Is this a bug?

And on that page it showed an identicon with the Parity Signer logo embedded inside it. Is this a bug too?

i.e. how does it know that it was generated with Parity Signer when I'm importing it using the recovery words approach instead of the QR scanner approach, or is it showing the Parity Signer logo embedded in the identicon because it's already in my Fether account list since it was previously imported using the QR code?

<img width="360" alt="screen shot 2019-01-10 at 5 30 38 pm" src="https://user-images.githubusercontent.com/6226175/51002670-257a0500-1588-11e9-9615-6907c6ca62e3.png">

So then I gave it a different name "test-signer2-again", and clicked "Next" where it prompted me for a password, which I entered as shown below:

<img width="361" alt="screen shot 2019-01-11 at 9 31 23 am" src="https://user-images.githubusercontent.com/6226175/51003041-5ad32280-1589-11e9-915e-83c5facdf7ef.png">

It then returned me to the Accounts page, but now I have both the "test-signer2" account (WITH the Parity Signer logo embedded), and the "test-signer2-again" account (WITHOUT the Parity Signer logo embedded), and they both have the same Ethereum address

<img width="361" alt="screen shot 2019-01-11 at 9 32 08 am" src="https://user-images.githubusercontent.com/6226175/51003428-dc778000-158a-11e9-85a2-e732b9f5f9e1.png">

When I went to the "test-signer2" account page it still showed both ETH and THIBCoin tabs,
but when I went to the "test-signer2-again" account page it only showed the ETH tab, and I had to add the THIBCoin tab again.

If I check Dev Tools > Application > Local Storage, it hasn't updated the following list of Parity Signer accounts:
* Key: `localforage/__paritylight::paritySignerAccounts`
* Value: 
```
[{"address":"0x006e27b6a72e1f34c626762f3c4761547aff1421","name":"test-signer"},{"address":"0x000ec8c79dac74d13b7e4743ae52f78d7186ba6d","name":"test-signer2"},{"address":"0x00d7bb9dbd761fdf00d78f74710ad6bdcaed8d48","name":"test-signer-mainnet"},{"address":"0x00ca490242abb36cfad5697bafa69fbd1824389d","name":"test4-ropsten"}]
```

And the following entry for the account hasn't changed:
* Key: `localforage/__paritylight::tokens::0x000ec8c79dac74d13b7e4743ae52f78d7186ba6d::kovan`
* Value: 
```
{"ETH":{"address":"ETH","decimals":18,"logo":"data:image/png;base64,iVB...YII=","name":"Ether","symbol":"ETH"},"0x4A6e6C3868A279e1D9047B42C3fB356FF4680003":{"address":"0x4A6e6C3868A279e1D9047B42C3fB356FF4680003","decimals":18,"name":"THIBCoin","symbol":"TIB","logo":"http://pix.toile-libre.org/upload/original/1525765045.jpg"}}
```

So... my thoughts are:

i) If the user tries to import the same Parity Signer account into Fether using "Recover from Recovery Words" when they had already previously imported that account into Fether from Parity Signer and that account is already shown in their account list, then it should display `Account 0x... already listed`), instead of creating a duplicate account. This will require making updates to apply to "Recover from Parity Signer" similar changes to those that were made in this PR https://github.com/paritytech/fether/pull/327.

ii) If the user does not already have a specific Parity Signer account in their Fether account list and they choose to import it using "Recover from Recovery Words" instead of "Recover from Parity Signer", do we still consider it as being "Parity Signer"-originated and show the Parity Signer logo embedded in the account identicon? I don't think we should.

I think we should somehow communicate to the user that if they don't import an account that they created with Parity Signer using "Recover from Parity Signer" that it won't have the Parity Signer logo embedded in the account identicon...

I also think that if they forget to import an account that they created with Parity Signer using "Recover from Parity Signer", they may find themselves upset that they don't have the Parity Signer logo embedded in the account identicon. But there's currently no way to delete their account from Fether and then re-import it again but using "Recover from Parity Signer" so that they get the Parity Signer logo embedded in the account identicon. So I think we need a Fether feature to Delete Account just like we have in Parity Signer (once we've got a Fether menu https://github.com/paritytech/fether/issues/264#issue-383304369)

ltfschoen commented on packages/fether-react/package.json:
```diff
@@ -35,14 +35,18 @@
-    "@parity/api": "^3.0.11",
+    "@parity/abi": "^3.0.11",
+    "@parity/api": "^3.0.25",
     "@parity/contracts": "^3.0.11",
```
FYI, I just tried using 3.0.25 [here](https://github.com/paritytech/fether/pull/352#issuecomment-453310120) and I got this error again `error An unexpected error occurred: "Failed to replace env in config: ${NPM_TOKEN}".`

ltfschoen replied:
LGTM

axelchalon commented on packages/fether-react/src/Send/SignedTxSummary/SignedTxSummary.js:
```diff
@@ -0,0 +1,135 @@
+                      <form className='send-form' onSubmit={handleSubmit}>
+                        <fieldset className='form_fields'>
+                          <Field
+                            className='form_field_amount'
```
fixed in the merge commit

axelchalon commented:
Thanks for the write-up Luke,

> could we also show what network it is associated with just like we do in Parity Signer and colour-code it?

`@parity/qr-signer` returns us the address and chainId when scanning the account, so this can be done. However I think it might be easier to simply show the Signer account in the accounts list if we're connected to the right chain (we might want to show a warning at one point if we import a Signer account for a chainId other than the chain we're currently connected to -- otherwise the user might be surprised not to see their imported signer account show up in the accounts list after completing the import). I completely forgot to scope the Signer accounts to their chainId; fixed it in the latest commit. 

> 6) Error with an unknown cause

also happened to me. I haven't checked if it prevents us form scanning?

> 7) Unable to Send Transaction using Fether from an account imported from Parity Signer

I don't think I can reproduce; or the issue specifically linked to transferring across chains? I'd be surprised because the account addresses all use the same format afaik?

> 8) Gap between Identicon and Account name

it is actually related to this PR; added some quickfix for the margin.

axelchalon commented on packages/fether-ui/package.json:
```diff
@@ -31,6 +31,7 @@
     "start": "yarn build --watch"
   },
   "dependencies": {
+    "babel-plugin-inline-import-data-uri": "^1.0.1",
```
`import signerIcon from './signerIcon.png';` in this PR's changes to `fether-ui/src/AccountCard/Avatar/Avatar.js`

https://github.com/paritytech/fether/pull/336/files/94f2db7a920ddd15cc42db04d331d460836b438e#diff-e393dc70fffc9ab07f0ea3e44e712791R10

Wanted to find an easy way to be able to use images in fether-ui (currently fether-ui doesn't have any images)

axelchalon commented:
TODO:
- [ ] Fix refeshing
- [ ] Fix webcam bars (tracking https://github.com/parity-js/qr-signer/pull/39)
- [ ] Fix occasional bug when submitting the TxForm `"Cannot read toNumber of null"` because somehow `gasPrice` is sometimes undefined when submitting the TxForm, not sure why

Tbaut commented on packages/fether-react/src/Send/TxQrCode/TxQrCode.js:
```diff
@@ -0,0 +1,96 @@
+                  account={address}
+                  rlp={getRlp()}
+                />
+              </Card>
```
We should include the buttons in the card.

Tbaut commented on packages/fether-react/src/Send/ScanSignedTx/ScanSignedTx.js:
```diff
@@ -0,0 +1,90 @@
+                  onScan={this.onScanSignedTx}
+                  label='Please show the QR code of the signed transaction on the webcam'
+                />
+              </Card>
```
We should include the buttons in the card here as well

ltfschoen commented:
I just pulled the latest changes from this branch `git pull --rebase origin ac-signer`.
I then started it up `yarn; yarn start`.
I then went to the recover from QR code page and it displayed the following:

<img width="1133" alt="screen shot 2019-01-12 at 11 32 15 am" src="https://user-images.githubusercontent.com/6226175/51066385-e9b16f00-165d-11e9-9a1f-b0555ec19760.png">

I killed it and restarted again with `yarn; yarn start`, but it still occurred.
So I added `console.log('Unknown error: ', e);` in the switch statement where it's printing out "Unknown error." and found that the error name is `NotReadableError`

<img width="1138" alt="screen shot 2019-01-12 at 11 40 14 am" src="https://user-images.githubusercontent.com/6226175/51066613-6264fb00-165f-11e9-835a-dbc2bb05e856.png">

I think it's because the FaceTime HD Camera (built-in) webcam on my MacBook Pro encountered an error, because when I try to create a New Movie Recording in QuickTime it says it cannot use the webcam. Hopefully I just have to restart the computer.

So I think we need to update the `default` case in the switch statement to perform a `console.log` for unknown error reporting. And to add an additional case with a message similar to the following:
```
case 'NotReadableError':
  errorMessage = 'Webcam hardware error. Try restarting your computer';
  break;
```

ltfschoen commented:
10. Backup to JSON file of an account that was created with Parity Signer but imported into Fether using "Recover from Seed Phrase"

I just created an account using Parity Signer.

But I wanted to have the option of either:
* Recovering the account using the recovery words (seed phrase), OR
* Recovering the account from a JSON file backup.

Now I know that if I imported the account into Fether using "Recover from Parity Signer" then if I went to that account's page in Fether it wouldn't show the option to "Backup Account" so I could save it to a JSON file as an additional backup option

So instead of importing the account into Fether using "Recover from Parity Signer", I chose to "Recover from Seed Phrase", and it successfully imported the account (but it does not show the Parity Signer logo embedded into the identicon).

In the Fether, on the Account List I clicked the account to take me to the Account page, where it displayed the "Backup Account" page, and I got all excited! So I downloaded the JSON file backup for the account.

So to I assumed that I could import the JSON file since it's worked in the past, but I wanted to double check. So I navigated to `cd ~/Library/Application Support/io.parity.ethereum/keys/kovan`, and viewed the accounts by creation date `ls -al`, and temporarily moved the most recently created one into a ./temp subdirectory. i.e. `mkdir -p ./temp && UTC--2019-01-12T00-18-13Z--6fa...5fc ./temp/UTC--2019-01-12T00-18-13Z--6fa...5fc`. Then it was no longer listed in the Fether account list.

Then in Fether I went to "+ > Import account > Recover from JSON Keyfile" and chose the JSON file that was downloaded. It successfully imported the account.

Why don't users who import an account they created with Parity Signer using "Recover from Parity Signer" get the same privilege as users who import it with "Recover from Seed Phrase" instead, such that they are able to backup the account to a JSON file if they want to?

ltfschoen commented:
## 12. Broken 'Details' button

See this issue that relates to this PR https://github.com/paritytech/fether/issues/354

Tbaut commented:
>Why don't users who import an account they created with Parity Signer using "Recover from Parity Signer"...

This is the whole point of Parity Signer, nobody else than the phone has the private key. The JSON is the (encrypted) private key. Fether doesn't have it for such accounts.

axelchalon commented:
it looks like the reset/refreshing bug is fixed thanks to Amaury's refactoring suggestion :tada:

ltfschoen commented:
I've gone through all my previous comments and checked to see if they've been fixed (able to close my dot points 1, 2, 4, 5, 8, 10). Here are the results using Fether generated Parity Ethereum node using latest from this branch:

* [X] Tested that can send ETH from an account created **without** Parity Signer on Kovan network to another account account that was created using Parity Signer on Mainnet or Ropsten network

* [X] Tried importing a Parity Signer account that had already been imported both using Recover from Parity Signer and using Recover from Seed Phrase and the functionality works for both

* [X] If I move all the keys out of this folder ~/Library/Application Support/io.parity.ethereum/keys/kovan, all the Parity Signer accounts that I imported are the only ones shown in the Fether account list (since they are only stored in Local Storage)

* [X] Local storage values of an account imported from Parity Signer has the correct network name since they are being stored using a `chainId` associated with the account. And if I add tokens using a Kovan network account it shows `::kovan` in the Local Storage entry key 

i.e.
```
[{"address":"0x00d7bb9dbd761fdf00d78f74710ad6bdcaed8d48","name":"test3-mainnet","chainId":1},{"address":"0x00ca490242abb36cfad5697bafa69fbd1824389d","name":"test4-ropsten","chainId":3},{"address":"0x000ec8c79dac74d13b7e4743ae52f78d7186ba6d","name":"test2-kovan","chainId":42}]
```

* [X] I was able to successully send a transaction from an Kovan network account imported from Parity Signer using QR codes to sign the tx to a non-Parity Signer generated account

* [X] Tested that if the user does not already have a specific Parity Signer account in their Fether account list and they choose to import it using "Recover from Recovery Words" instead of "Recover from Parity Signer", it no longer shows the Parity Signer logo embedded in the account identicon since it's not considered "Parity Signer"-originated. They are able to Backup Account to JSON file
 
* [X] "9) Handling a webcam hardware error" appears to have been addressed

* [X] "11) TypeError: Invalid hex string, when accidently scanned the sending accounts QR code instead of the signed transaction's QR code" is resolved, as it now displays "The QR code doesn't appear to be a valid transaction"

* [X] "12) Broken 'Details' button" is fixed

* [.] Since I am no longer able to list non-"Kovan" network accounts in the Fether account list, I am no longer able to encounter the issue that occurred when trying to transfer a value across chains that I mentioned in "7) Unable to Send Transaction using Fether from an account imported from Parity Signer" 

* [ ] Tested that can send ETH from account created **with** Parity Signer on Ropsten network to another account created with Parity Signer on Mainnet. After entering amount and recipient address, it showed QR code and when scanned with Parity Signer it showed error in Parity Signer (where account 00ca49...69f is the Parity Signer on Ropsten network account):

![screenshot_20190115-195453_parity signer](https://user-images.githubusercontent.com/6226175/51169674-9f9be800-1900-11e9-8229-c2a3186b28c6.jpg)

* [ ] After trying to import an account that already exists the notification "Account xxx already listed" should be centred

<img width="358" alt="screen shot 2019-01-15 at 8 10 16 pm" src="https://user-images.githubusercontent.com/6226175/51170069-9bbc9580-1901-11e9-8913-c018d194d285.png">

* [ ] When I try to create any new account in Parity Signer it doesn't show any Recovery Words anymore, the area is blank. I tried restarting Parity Signer but the problem still persists. I haven't tried reinstalling Parity Signer yet (as I'd probably have to add all the accounts again)

![screenshot_20190115-201641_parity signer](https://user-images.githubusercontent.com/6226175/51170815-76308b80-1903-11e9-996a-6ff1a8abd616.jpg)

* [ ] If I remove a Parity Signer "Mainnet" account from the account list in Fether by updating Local Storage by removing the key/value `localforage/__paritylight::tokens::<ACCOUNT_ID>` for the account address entry and removing the account from the array under the key `localforage/__paritylight::paritySignerAccounts`. And then in Fether Recover from Parity Signer with QR code, it adds it again to the array in Local Storage `[...., {"address":"0x00d7bb9dbd761fdf00d78f74710ad6bdcaed8d48","name":"test3-signer-mainnet","chainId":1}]`, but it doesn't display in the Fether account list again, even after refreshing.

So I cleared all the key/value pairs from Local Storage and refreshed Fether and it displayed the T&Cs page, then when I pressed Accept, it showed the following error:
<img width="359" alt="screen shot 2019-01-15 at 8 41 14 pm" src="https://user-images.githubusercontent.com/6226175/51171831-ffe15880-1905-11e9-8aba-166d02cddbca.png">

When I refreshed again it went to T&Cs page again and clicked Accept it and it went to the account list page ok. So now I didn't have any accounts listed in the account list page.

I tried to import the Parity Signer "Mainnet" account again using Recover from Parity Signer, but although it did add it to Local Storage as shown below, it didn't appear in the account list page again, perhaps because it's now stored with `chainID: 1` (mainnet)
<img width="1234" alt="screen shot 2019-01-15 at 8 52 30 pm" src="https://user-images.githubusercontent.com/6226175/51172533-88acc400-1907-11e9-8f7d-b054cd4e5b5f.png">

So I decided to start a Parity Ethereum node on "Mainnet" separately `./target/release/parity --chain ethereum --light` and then restart Fether to see if only that Parity Signer "Mainnet" account that I had imported would be shown in the account list, but when Fether loaded it showed the T&Cs page, then when I clicked "Accept" it showed error below.

<img width="1095" alt="screen shot 2019-01-15 at 8 56 18 pm" src="https://user-images.githubusercontent.com/6226175/51172835-391ac800-1908-11e9-905a-e0636324a717.png">

<img width="357" alt="screen shot 2019-01-15 at 8 56 38 pm" src="https://user-images.githubusercontent.com/6226175/51172843-3ae48b80-1908-11e9-9367-9c2a8350ec63.png">

This is the same as the error that occurs if you try to run a 2nd Parity Ethereum instance. See https://github.com/paritytech/fether/issues/333

If I clicked the "X" button it went back to T&Cs page, and this time when I clicked "Accept" the UI went blank, even after refreshing, and in the Dev Tools console it displayed:

<img width="1237" alt="screen shot 2019-01-15 at 9 00 17 pm" src="https://user-images.githubusercontent.com/6226175/51173037-b21a1f80-1908-11e9-9d0d-e3d17d239ebe.png">

So I went back to starting a Parity Ethereum node on "Kovan" separately `./target/release/parity --chain kovan --light` and then restart Fether to try and get it working again, but when Fether loaded it showed a blank screen, even after refreshing, and it showed the same `chainId` error in Dev Tools console.

So as a last resort I just started Fether by itself so it would generate the Parity Ethereum node, and that worked as before.

**So it looks like Fether is crashing now when you run a separate Parity Ethereum Light node even with the Kovan network before starting Fether. Perhaps it's trying to create a 2nd node?**. Note that I tried starting it with `Parity Ethereum
  version Parity-Ethereum/v2.1.10-stable-b47e064f8-20181213/x86_64-macos/rustc1.31.0`, which worked previously in 'master' branch.

But it is correctly not listing accounts that were created for non-Kovan networks with Parity Signer that the user has imported into Fether (since Fether generates only a Kovan network node at the moment)

* [ ] regarding "5) Importing Account in Fether using the passphrase that was generated when I created an in Parity Signer (Continued)". 
How are we communicating/warning the user that if they don't import an account that they created with Parity Signer using "Recover from Parity Signer" that it won't have the Parity Signer logo embedded in the account identicon?
And how are we communicating/warning the user that if they don't import an account that they created with Parity Signer using "Recover using Seed Phrase" that it won't give them the option to Backup Account to a JSON file (so they know they have that option available too)?

* [ ] If the user forgets to import an account that they created with Parity Signer using "Recover from Parity Signer", they may find themselves upset that they don't have the Parity Signer logo embedded in the account identicon and want to re-import it again using "Recover from Seed Phrase" instead, but we don't have a Delete Account option in Fether yet.
Or if they import an account that they created with Parity Signer using "Recover from Parity Signer", but they wanted the ability to Backup Account, there isn't a Delete Account option in Fether to allow them to re-import it using "Recover from Seed Phrase" instead so they have that option available.
So I think we need a Fether feature to Delete Account just like we have in Parity Signer (once we've got a Fether menu #264 (comment)). Should we create an New Issue for this?

* [ ] If I try to send 0.0005 ETH from a Parity Signer imported account in Fether to another account, it prompts me to scan the QR code of the transaction on Parity Signer, which I did, and in Parity Signer it shows "Transaction Details" of 0.0005 ETH from 0x000ec8... to 0.005072... (both on Kovan). If I then click "Next Step" in Fether, it asks me to show the QR code of the signed transaction on the webcam, which I would get by clicking "Sign Transaction" in Parity Signer. 
**But instead I decide to scan a QR code signature from a "previous" transaction of only 0.001 ETH**.
In Fether it then takes me to the following screen where it still shows I want to send 0.0005 ETH, and a "Send" button.

<img width="358" alt="screen shot 2019-01-15 at 10 40 27 pm" src="https://user-images.githubusercontent.com/6226175/51178479-c8c77300-1916-11e9-9b70-a122e50311a7.png">

If I click "Send" it signs it and shows "Sending your transaction...". It's taking a long time and I still haven't received any block confirmations...
But anyway, the purpose was just to check if 0.001 ETH or 0.0005 ETH gets transferred at https://blockscout.com/eth/kovan/address/0x005072fb985cc64d8a1ce00ece26279f73c12f8a/transactions. What amount do you think should be transferred and if the signature QR that I used was associated with a previous transaction to a different recipient would it send it to them, or the recipient that was entered in the input field in Fether?

If I wanted to repeat the same transaction amount multiple times to the same recipient could I just ignore scanning the QR code shown by Fether and skip to the next step and then just show the same screenshot of the original signature QR code that I saved to my device?

* [ ] regarding "6) Error with an unknown cause", should we create a New Issue? I'm not sure how to reproduce this again though...

Tbaut commented:
I see no issue left in any of the points raised @ltfschoen (except the last one, "6) Error with an unknown cause", if you can reproduce.). 

The centring of the error message was known to me but doesn't apply to this PR.

If the code is ok, I'm good to merge.

ltfschoen commented:
> I see no issue left in any of the points raised @ltfschoen (except the last one, "6) Error with an unknown cause", if you can reproduce.).
> 
> The centring of the error message was known to me but doesn't apply to this PR.
> 
> If the code is ok, I'm good to merge.

Yes I don't see any issues merging it. It'd be good to get some quick feedback on the "unchecked" checkboxes though (i.e. whether a separate issue should be raised after this PR is merged, or whether I've potentially misunderstood something, or ..)

Tbaut commented:
Sure, here's my feedback:
> And how are we communicating/warning the user that if they don't import an account that they created with Parity Signer using "Recover using Seed Phrase" that it won't give them the option to Backup Account to a JSON file (so they know they have that option available too)?

I think this should be understood separately from Fether and as part of the signer workflow. A user installing and using Signer will know why it does this. Recovery is not something users do every day.

>  Importing Account in Fether using the passphrase that was generated when I created an in Parity Signer

if you import the account with the recovery phrase, then it'll be in Fether, and since the recovery passphrase =  private key, Signer has no role to play anymore. It's normal not to have the Signer logo.

> If I wanted to repeat the same transaction amount multiple times to the same recipient could I just ignore scanning the QR code shown by Fether and skip to the next step and then just show the same screenshot of the original signature QR code that I saved to my device?

This wouldn't work and this is what transaction nonces are made for, to avoid replay transactions.
